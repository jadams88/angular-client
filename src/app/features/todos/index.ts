export { Todo } from './todos.model';
export { TodoModule } from './todos.module';
export { TodoService } from './services/todos.service';
export * from './actions/todos.actions';
