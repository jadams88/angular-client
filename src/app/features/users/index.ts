export { User } from './user.model';
export * from './user.actions';
export { usersReducer, selectUser } from './user.reducer';
export { UserService } from './user.service';
