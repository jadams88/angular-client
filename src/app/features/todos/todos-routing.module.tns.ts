import { NgModule } from '@angular/core';
import { Routes } from '@angular/router';
import { TodosComponent } from './components/todos.component';
import { NativeScriptRouterModule } from 'nativescript-angular/router';

const routes: Routes = [
  { path: '/:id', component: TodosComponent },
  { path: '', pathMatch: 'full', component: TodosComponent }
];

@NgModule({
  imports: [NativeScriptRouterModule.forChild(routes)],
  exports: [NativeScriptRouterModule]
})
export class TodoRoutingModule {}
