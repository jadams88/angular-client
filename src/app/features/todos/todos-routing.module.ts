import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { TodosComponent } from './components/todos.component';

const routes: Routes = [{ path: '', pathMatch: 'full', component: TodosComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class TodoRoutingModule {}
