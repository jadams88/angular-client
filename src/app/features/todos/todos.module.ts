import { NgModule } from '@angular/core';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { SharedModule, AppMaterialModule } from '@app/shared';
import { todosReducers, TodoState } from './todos.reducer';
import { TodoEffects } from './todos.effects';
import { TodoService } from './todos.service';
import { TodosComponent } from './components/todos.component';
import { TodoListComponent } from './components/todo-list/todo-list.component';
import { TodoDetailComponent } from './components/todo-detail/todo-detail.component';
import { TodoRoutingModule } from './todos-routing.module';

const COMPONENTS = [TodosComponent, TodoListComponent, TodoDetailComponent];

@NgModule({
  imports: [
    SharedModule,
    AppMaterialModule,
    TodoRoutingModule,
    StoreModule.forFeature<TodoState>('todos', todosReducers),
    EffectsModule.forFeature([TodoEffects])
  ],
  providers: [TodoService],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class TodoModule {}
