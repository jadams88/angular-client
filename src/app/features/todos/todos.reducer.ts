import { Todo } from './todos.model';
import { ActionReducerMap, createFeatureSelector, createSelector } from '@ngrx/store';
import { TodoActionTypes, TodoActionUnion } from './todos.actions';

const COMPARATOR = 'id';

export interface TodoState {
  todos: Todo[];
}

export function todosReducer(state: Todo[] = [], action: TodoActionUnion): Todo[] {
  switch (action.type) {
    case TodoActionTypes.LoadSuccess:
      return <Todo[]>action.payload;

    case TodoActionTypes.CreateSuccess:
      return <Todo[]>[...state, action.payload];

    case TodoActionTypes.UpdateSuccess:
      return <Todo[]>state.map(todo => {
        return todo[COMPARATOR] === action.payload[COMPARATOR]
          ? Object.assign({}, todo, action.payload)
          : todo;
      });

    case TodoActionTypes.Delete:
      return <Todo[]>state.filter(todo => {
        return todo[COMPARATOR] !== action.payload[COMPARATOR];
      });

    default:
      return state;
  }
}

// a map of the todos reducers
export const todosReducers: ActionReducerMap<TodoState> = {
  todos: todosReducer
};

// Select the top level 'todos' state.
export const selectTodoState = createFeatureSelector<TodoState>('todos');

// todo sub state selectors
export const selectTodos = createSelector(
  selectTodoState,
  (state: TodoState) => state.todos
);
