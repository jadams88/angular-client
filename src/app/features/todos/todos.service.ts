import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { ApiService } from '@app/core';
import { Todo } from './todos.model';
import { AppState } from '@app/store/reducers';
import { selectTodos } from './todos.reducer';
import { UpdateTodo, CreateTodo } from '@app/features/todos/todos.actions';

@Injectable()
export class TodoService {
  public todo$: Observable<Todo[]>;

  constructor(private store: Store<AppState>, private api: ApiService) {
    this.todo$ = this.store.pipe(select(selectTodos));
  }

  public loadTodos(): Observable<Todo[]> {
    return this.api.get<Todo[]>('todos');
  }

  public getTodo(id: string): Observable<Todo[]> {
    return this.api.get<Todo[]>(`todos/${id}`);
  }

  public saveTodo(todo: any): void {
    const action = todo.id ? new UpdateTodo(todo) : new CreateTodo(todo);
    this.store.dispatch(action);
  }

  public createTodo(todo: Todo): Observable<Todo> {
    return this.api.post<Todo>('todos', todo);
  }

  public updateTodo(todo: Todo): Observable<Todo> {
    return this.api.put<Todo>('todos', todo);
  }

  public deleteTodo(todo: Todo): Observable<Todo> {
    return this.api.delete<Todo>('todos', todo);
  }
}
