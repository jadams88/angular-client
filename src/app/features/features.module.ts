import { NgModule } from '@angular/core';
import { UsersModule } from './users/users.module';

/**
 * The Features module is only a convenience module to barrel all modules
 * That are NOT to be lazy loaded. This just helps to keep the AppModule clean off
 * all the modules
 */
@NgModule({
  imports: [UsersModule],
  exports: [UsersModule]
})
export class FeaturesModule {}
