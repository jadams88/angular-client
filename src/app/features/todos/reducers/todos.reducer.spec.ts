// A reducer is just a pure function, it takes inputs and return a value when called
// Because it is pure function we do not need to configure any of angular testing
// module because it does not need of angular configuration to operate.
// We can also just use snapshot testing to test the output of the function
// and not need to use assertion testing
import { todosReducer, TodosEntityState } from './todos.reducer';
import { Todo } from '@app/features/todos';
import {
  LoadTodosSuccess,
  CreateTodoSuccess,
  UpdateTodoSuccess,
  DeleteTodoSuccess
} from '../actions/todos.actions';
import { EntityAdapter, createEntityAdapter, EntityState, Dictionary } from '@ngrx/entity';
import { TodoFilterStatus } from '../models/todos.model';

describe('TodoReducer', () => {
  let adapter: EntityAdapter<Todo>;

  const todo: Todo = {
    id: '1',
    user: '1',
    title: 'some title',
    description: 'some description',
    completed: true
  };
  const initialState: TodosEntityState = {
    ids: [todo.id],
    entities: {
      [todo.id]: todo
    },
    selectedTodoId: null,
    allTodoFilter: TodoFilterStatus.InCompleted,
    allTodoFilterString: null
  };

  beforeEach(() => {
    adapter = createEntityAdapter<Todo>();
  });

  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {} as any;
      const result = todosReducer(undefined, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('LoadSuccess', () => {
    it('should add the todos to the todo state', () => {
      const todos: Todo[] = [
        {
          id: '1',
          user: '1',
          title: 'some title',
          description: 'some description',
          completed: true
        },
        {
          id: '2',
          user: '1',
          title: 'another title',
          description: 'another description',
          completed: false
        }
      ];

      const action = new LoadTodosSuccess(todos);
      const result = todosReducer(undefined, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('CreateSuccess', () => {
    it('should add a todo to the todo state', () => {
      const newTodo = {
        id: '2',
        user: '1',
        title: 'another title',
        description: 'another description',
        completed: false
      };

      const action = new CreateTodoSuccess(newTodo);
      const result = todosReducer(initialState, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('UpdateSuccess', () => {
    it('should update the todo state', () => {
      const todo1 = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: false
      };
      const todo2 = { ...todo1, id: '2', title: 'anotherTodo' };

      const state: TodosEntityState = {
        ids: [todo1.id, todo2.id],
        entities: {
          [todo1.id]: todo1,
          [todo2.id]: todo2
        },
        selectedTodoId: null,
        allTodoFilter: TodoFilterStatus.InCompleted,
        allTodoFilterString: null
      };

      const updateTodo = {
        id: '2',
        completed: true
      } as Todo;

      const action = new UpdateTodoSuccess(updateTodo);
      const result = todosReducer(state, action);
      expect(result).toMatchSnapshot();
    });
  });

  describe('DeleteSuccess', () => {
    it('should delete the todo from the sate', () => {
      const todo1 = {
        id: '1',
        user: '1',
        title: 'some title',
        description: 'some description',
        completed: true
      };
      const todo2 = { ...todo1, id: '2', title: 'anotherTodo' };

      const state: TodosEntityState = {
        ids: [todo1.id, todo2.id],
        entities: {
          [todo1.id]: todo1,
          [todo2.id]: todo2
        },
        selectedTodoId: null,
        allTodoFilter: TodoFilterStatus.InCompleted,
        allTodoFilterString: null
      };

      const todoToDelete = {
        id: '2'
      } as Todo;

      const action = new DeleteTodoSuccess(todoToDelete);
      const result = todosReducer(state, action);
      expect(result).toMatchSnapshot();
    });
  });
});
