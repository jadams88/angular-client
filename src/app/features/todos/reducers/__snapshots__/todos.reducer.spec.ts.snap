// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoReducer CreateSuccess should add a todo to the todo state 1`] = `
Array [
  Object {
    "completed": true,
    "description": "some description",
    "id": 1,
    "title": "some title",
  },
  Object {
    "completed": false,
    "description": "another description",
    "id": 2,
    "title": "another title",
  },
]
`;

exports[`TodoReducer DeleteSuccess should delete the todo from the sate 1`] = `
Array [
  Object {
    "completed": true,
    "description": "some description",
    "id": 1,
    "title": "some title",
  },
  Object {
    "completed": false,
    "description": "another description",
    "id": 2,
    "title": "another title",
  },
]
`;

exports[`TodoReducer LoadSuccess should add the todos to the todo state 1`] = `
Array [
  Object {
    "completed": true,
    "description": "some description",
    "id": 1,
    "title": "some title",
  },
  Object {
    "completed": false,
    "description": "another description",
    "id": 2,
    "title": "another title",
  },
]
`;

exports[`TodoReducer UpdateSuccess should update the todo state 1`] = `
Array [
  Object {
    "completed": true,
    "description": "some description",
    "id": 1,
    "title": "some title",
  },
  Object {
    "completed": true,
    "description": "another description",
    "id": 2,
    "title": "another title",
  },
]
`;

exports[`TodoReducer undefined action should return the default state 1`] = `Array []`;
